### PR TITLE
agent db: make rejecting ooo samples configurable

### PR DESCRIPTION
**Pull request by rabenhorst**
Add `storage.agent.reject-out-of-order-samples` flag and make rejecting out-of-order samples in agent's db configurable. This allows disabling remote writing out-of-order samples from agent, which was introduced in https://github.com/prometheus/prometheus/pull/12897.

### CODE REVIEW

**Comment by ArthurSens** [comment]
Maybe I'm missing something, but isn't it already possible to disable OOO ingestion with the configuration file?

From my understanding if `out_of_order_time_window` is zero (the default), then OOO is disabled. See https://prometheus.io/docs/prometheus/latest/configuration/configuration/#tsdb

**Comment by rabenhorst** [comment]
> From my understanding if `out_of_order_time_window` is zero (the default), then OOO is disabled. See https://prometheus.io/docs/prometheus/latest/configuration/configuration/#tsdb

I don't see how this applies to the agent db, which, as I understand, is just the TSDB WAL.

**Comment by sdufel** [comment]
When agents ingest out-of-order samples and and remote write them to external stores, it can be extremely expensive on the other end.  We run a setup with prometheus agents writing to thanos receivers, and out of order samples generate severe amounts of load.

**Comment by ArthurSens** [comment]
> Maybe I miss something, but I don't see how this applies to the agent db, which, as I understand, is just the TSDB WAL.

Yeah, I'm reading the code now and it looks like the agent indeed doesn't take this into consideration. 

I think that creating yet another configuration option for out of order would be confusing though. Could we re-use the already existing `out_of_order_time_window` for that?

**Comment by rabenhorst** [comment]
> Yeah, I'm reading the code now and it looks like the agent indeed doesn't take this into consideration.
> 
> I think that creating yet another configuration option for out of order would be confusing though. What do you think about re-using the already existing `out_of_order_time_window` configuration option?

I think it makes sense and I changed it and added some test cases for edge cases. Is this what you had in mind? 

It might be confusing now that `out_of_order_time_window` is also applied to agent db.  What would be the best place to document it?

File: cmd/prometheus/main.go
Comment (ArthurSens) [inline cmd/prometheus/main.go:1183]: Hmmm, maybe we could remove this new line and set the out-of-order window inside `ToAgentOptions`?

Not sure if that's nitpicking or not to be honest, what do you think?

File: tsdb/agent/db.go
Comment (ArthurSens) [reply tsdb/agent/db.go:94]: I see that you also use this function in tests, but we could simply use `opts.OutOfOrderTimeWindow = <somevalue>` over there

File: tsdb/agent/db_test.go
Comment (ArthurSens) [inline tsdb/agent/db_test.go:908]: What do you think about changing `reject bool` to `error`, that could be either unset(nil) or `storage.ErrOutOfOrderSample` when declaring the test cases?

This way we could delete this if, just a little less code to read 😛

Comment (ArthurSens) [reply tsdb/agent/db_test.go:908]: We wouldn't be able to delete the if below, but that's fine I guess

**Review by ArthurSens — COMMENTED** [review]
It LGTM, just a few nitpicks

I'll try to ping maintainers to take a look at this as well

**Comment by ArthurSens** [comment]
I'm also realizing that changing the default behavior now (always accept OOO -> have to configure window to accept) might be a breaking change. Since both agent-mode is a feature flag and OOO is marked as experimental in our docs, I think the change should be fine, but let's see what others think

**Comment by ArthurSens** [comment]
> It might be confusing now that `out_of_order_time_window`, which is part of tsdb config, is also applied to agent db.  What would be the best place to document it?

Good point 😬, I don't remember if the configuration documentation is auto-generated... I'm 90% sure it's not. So it should be fine to just add a statement [here](https://github.com/prometheus/prometheus/blob/main/docs/configuration/configuration.md#:~:text=%23%20Configures%20how,of%2Dorder%20window.) clarifying that it is also applied for agent-mode?

File: cmd/prometheus/main.go
Comment (rabenhorst) [reply cmd/prometheus/main.go]: Done. Is that what you had in mind?

File: tsdb/agent/db_test.go
Comment (rabenhorst) [reply tsdb/agent/db_test.go]: Makes sense. Done!

**Comment by rabenhorst** [comment]
> It LGTM, just a few nitpicks
> 
> I'll try to ping maintainers to take a look at this as well

I addressed the nits and thx!

**Comment by roidelapluie** [comment]
cc @rfratto

**Comment by rabenhorst** [comment]
> Since we're reusing an existing Prometheus config field, should we update docs for `out_of_order_time_window` around https://prometheus.io/docs/prometheus/latest/configuration/configuration/#tsdb to mention this dual use-case?

I added it to `out_of_order_time_window` docs: https://github.com/prometheus/prometheus/blob/017b43ec69b6113c3e4788a07f3936b597b386af/docs/configuration/configuration.md?plain=1#L3805-L3807

WDYT?

**Review by ArthurSens — APPROVED** [review]
Since the next release process starts next week, I'd merge this even without @rfratto's approval 😬.

@rabenhorst, it seems like you forgot to sign your lastest commit. Could you fix that please? 🙂

**Comment by rabenhorst** [comment]
> Since the next release process starts next week, I'd merge this even without @rfratto's approval 😬.
> 
> @rabenhorst, it seems like you forgot to sign your lastest commit. Could you fix that please? 🙂

Done.

**Comment by ArthurSens** [comment]
Thanks @rabenhorst

**Review by rfratto — COMMENTED** [review]
Sorry for the post-merge review and holding this up longer than it should've been, but this looks good to me!

**Pull request merged**
